### ISSUE TITLE
LevenbergMarquardt speed

### ISSUE THREAD

**Issue by rmlopes** [itdxer/neupy#118]
First of all, congrats, I like very much the design of this library. It seems that there is something strange (perhaps buggy) going on with the LevenbergMarquardt algorithm, or either I am doing something wrong.  I am using the following to test it out:

```python
import theano
import pandas as pd
import numpy as np

from sklearn.preprocessing import robust_scale
from sklearn.cross_validation import train_test_split
from neupy import algorithms, layers

#Load the data
data = pd.DataFrame(np.array(range(10000)), dtype='float32')
target = pd.DataFrame([x**2 for x in data[0]], dtype='float32')

theano.config.floatX = 'float32'

#the main split
mask = (np.random.rand(len(data)) < 0.8)
traindata = data[mask]
traintarget = target[mask]
testdata = data[~mask]
testtarget = target[~mask]

#split training data set into validation
x_train, x_test, y_train, y_test = train_test_split(
    traindata.astype(np.float32),
    traintarget.astype(np.float32),
    train_size=(.8)
)

network = algorithms.LevenbergMarquardt(
    [
        layers.Input(1),
        layers.Relu(32),
        layers.Relu(16),
        layers.Sigmoid(1),
    ],
    error='mse',
    step=0.01,
    verbose=True,
    shuffle_data=True,
    #momentum=0.99,
    #nesterov=True,
)

#Train the netwrok with validation
network.train(x_train, y_train, x_test, y_test, epochs=100)

#predict new samples
p = network.predict(testdata)
e2 = (testtarget - p) ** 2
mse = np.mean(e2)
meanobs = np.mean(testtarget)
ssobs = (testtarget - meanobs) ** 2
print("MSE: ")
print(mse)
print("R2: ")
print(1 - (mse/np.mean(ssobs)))
```

Each training iteration takes about 50 seconds, while most of the algorithms take just a few ms.

**Comment by itdxer** [comment]
Hi,

I'm not sure that I understand your problem correctly. Could you please clarify it in more details?

Also, I've noticed that you're are trying to predict integer number whether output from the network always between 0 and 1. I suggest you scale target variable (to make value smaller) and use Relu as a continuos function.

**Comment by rmlopes** [comment]
This is just a dummy testbed to show the issue, forget about scaling and whatnot. My problem is that LM takes 50 seconds, when for instance SGD (and many others) take just a few milliseconds. This seems like over exaggerated, like something is wrong.

**Comment by itdxer** [comment]
ok, Now it's clear. Levenberg-Marquardt is not very fast algorithm. It involves jaccobian matrix computation which is computationally expensive and in addition algorithm needs to find inverse matrix for the N by N matrix, where N is the number of parameters in the matrix, in your case it's 624. Anyway, I think that 50 seconds is too long for this tasks. I'll try to investigate this issue. Thanks.

**Comment by rmlopes** [comment]
I have been trying other algorithms. The Hessian seems to be suffering from a similar issue. It is not as slow as LM, but it still taking around 16 seconds per iteration.

There is probably something common underlying both algorithms that is causing the bottleneck. Some matrix operation perhaps?

**Comment by itdxer** [comment]
That's good to know. Hessian suppose to be slower than Levenberg-Marquardt. So there is definitely some issue with Levenberg-Marquardt, but I'm not totally sure that there is an issue with Hessian. It involves second order partial derivative calculation and inverse for the NxN matrix. I'll check it as well.

Thank you.

**Comment by rmlopes** [comment]
It definitely looks like a problem with the matrix operations, and LM is even slower because it has the Jacobian matrix computation besides the inverse of NxN matrix (which is a common operation in both).

**Comment by itdxer** [comment]
It looks like that jacobian computation is super slow. For your example during the training procedure algorithm spends 99.4% of the time in the loop that calculates jacobian. Not sure that I can fix it. On my laptop it takes 11-12 seconds per epoch. The main issue with the number of samples in your training dataset. In each iteration we need to find the gradient for every data sample which takes a lot of time. Even matrix inversion works very fast compare to this operation.

I don't see the way to speed it up in Theano. Later I'll try to research some different ways to solve this problem.

**Comment by itdxer** [comment]
@rmlopes, Do you have a code that reproduces NaN values for the Hessian algorithms?

**Comment by rmlopes** [comment]
I only get that in the real data, which I cannot share. I've tried playing around with the synthetic function but I could not reproduce it.

**Comment by itdxer** [comment]
I understand it. I haven't been able to reproduce it either. I'll continue investigating this issue
